app/lib/service/email/backend.dart

```diff
@@ -0,0 +1,137 @@
+import '../../shared/email.dart';
+import 'models.dart';
+
+final _logger = Logger('email.backend');
+
+/// Sets the email backend service.
+void registerEmailBackend(EmailBackend backend) =>
+    ss.register(#_emailBackend, backend);
+
+/// The active email backend service.
+EmailBackend get emailBackend => ss.lookup(#_emailBackend) as EmailBackend;
+
+/// Represents the backend for the outgoing email queue.
+class EmailBackend {
+  final DatastoreDB _db;
+
+  EmailBackend(this._db);
+
+  /// Creates [OutgoingEmail] entities that can be stored alongside a transaction.
+  List<OutgoingEmail> prepareEntities(EmailMessage msg) {
+    return msg.recipients
+        .map(
+          (recipient) => OutgoingEmail.init(
+            fromEmail: msg.from.email,
+            recipientEmail: recipient.email,
+            subject: msg.subject,
+            bodyText: msg.bodyText,
+          ),
+        )
+        .toList();
+  }
+
+  /// Queries all [OutgoingEmail] objects and tries to send out the email,
+  /// deleting the entry after the email was sent successfully.
+  ///
+  /// The processing of the Datastore query will be stopped after [stopAfter]
+  /// duration has elapsed. This allows the periodic task to complete within
+  /// the planned time window.
+  ///
+  /// When [skipRecentEntries] is true, the newly created entries are
+  /// not sent immediately, to prevent race conditions and double-sending
+  /// of emails.
+  ///
+  /// Returns the number of successfully sent emails.
+  Future<int> trySendAllOutgoingEmails({
+    Duration? stopAfter,
+    bool skipRecentEntries = false,
+  }) async {
+    final sw = Stopwatch()..start();
+    final query = _db.query<OutgoingEmail>()..order('-created');
+    var successful = 0;
+    await for (final m in query.run()) {
+      if (stopAfter != null && sw.elapsed > stopAfter) break;
+      if (skipRecentEntries && m.age.inMinutes == 0) continue;
+      if (m.isDead) continue;
+      if (!m.mayAttemptNow) continue;
+      if (await trySendOutgoingEmail(m.uuid)) {
+        successful++;
+      }
+    }
+    return successful;
+  }
+
+  /// Tries to send emails with the given [ids]. The
+  /// [OutgoingEmail] entry will be deleted after the
+  /// email was sent successfully.
+  Future<void> trySendOutgoingEmails(Iterable<String> ids) async {
+    for (final id in ids) {
+      await trySendOutgoingEmail(id);
+    }
+  }
+
+  /// Tries to send email with the given [id]. The
+  /// [OutgoingEmail] entry will be deleted after the
+  /// email was sent successfully.
+  ///
+  /// Returns true if the email was sent successfully.
+  Future<bool> trySendOutgoingEmail(String id) async {
+    return await withRetryTransaction(_db, (tx) async {
+      final o = await tx.lookupOrNull<OutgoingEmail>(
+          _db.emptyKey.append(OutgoingEmail, id: id));
+      if (o == null || o.isDead || !o.mayAttemptNow) {
+        return false;
+      }
+      try {
+        await emailSender.sendMessage(EmailMessage(
```

Comment:
Maybe we should generate a `Message-ID` header.
See: https://www.rfc-editor.org/rfc/rfc5322#section-3.6.4

And guidelines from: https://www.jwz.org/doc/mid.html

That way, if we resend it, the server should be able to deduplicate the message (or at-least group them in the same conversation).

Reply:
Done.